### PR TITLE
Bug 1335331: added ability to change email address

### PR DESCRIPTION
**Pull request by karabellyj**
@mathjazz r?

### CODE REVIEW

File: pontoon/contributors/static/css/contributor.css [hunk 1]
```diff
@@ -30,60 +30,37 @@ a.avatar:hover .desc ~ img {
...
+#profile-form .field input {
   color: #FFFFFF;
-  font-size: 48px;
-  font-weight: 300;
-  padding: 4px 50px;
-  text-align: center;
-  width: 100%;
+  background: #373C42;
+  border: 1px solid #4F5965;
+  border-radius: 5px;
```
Comment (mathjazz) [inline pontoon/contributors/static/css/contributor.css]: We should use the exact same styling as in other input fields, e.g. in the Filter projects field:
https://pontoon.mozilla.org/projects/

That is:
```
background: #333941;
border: 1px solid #4d5967;
border-radius: 3px;
```

File: pontoon/contributors/templates/contributors/profile.html [hunk 2]
```diff
@@ -24,17 +24,6 @@
...
-    {% if my_profile %}
-      <div class="container">
-        <input type="text" placeholder="What's your name?" value="{{ contributor.first_name }}" maxlength="30">
-        <div class="submit fa fa-arrow-circle-right"></div>
-      </div>
-    {% else %}
-      <div class="uneditable">{{ contributor.first_name }}</div>
-    {% endif %}
-  </h2>
-
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/profile.html]: We should keep the contributors's name displayed, but in read only mode.

File: pontoon/contributors/templates/contributors/settings.html [hunk 3]
```diff
@@ -21,12 +21,6 @@
     <img class="rounded" src="{{ user.gravatar_url(200) }}">
   </a>
 
-  <h2 id="username">
-    <div class="container">
-      <input type="text" placeholder="What's your name?" value="{{ user.first_name }}" maxlength="30">
-      <div class="submit fa fa-arrow-circle-right"></div>
-    </div>
-  </h2>
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html]: We should keep the contributors's name displayed, but in read only mode.

File: pontoon/contributors/templates/contributors/settings.html [hunk 4]
```diff
@@ -39,6 +33,21 @@ <h2 id="username">
...
+      </div>
+      <div class="field">
+          <label for="profile-email">Email</label>
+          <input type="email" id="profile-email" placeholder="email" value="{{ user.email }}">
+      </div>
+      <div class="controls">
+        <button class="submit button active">Save</button>
+      </div>
+  </div>
+
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html]: We should only have one form and one save button on the /settings page.

File: pontoon/contributors/static/css/contributor.css [hunk 5]
```diff
@@ -33,57 +33,42 @@ a.avatar:hover .desc ~ img {
...
   -moz-box-sizing: border-box;
   box-sizing: border-box;
 }
 
-#username .container input:hover,
-#username .container input:focus {
+#profile-form .field input:hover,
+#profile-form .field input:focus {
   border-color: #888888;
 }
```
Comment (mathjazz) [inline pontoon/contributors/static/css/contributor.css]: Please remove that rule. We don't use such effect on regular form fields.

File: pontoon/contributors/templates/contributors/settings.html [hunk 6]
```diff
@@ -39,6 +34,18 @@ <h2 id="username">
 
 {% block bottom %}
 <section id="main">
+  <div id="profile-form">
+      <h3>Personal infomation</h3>
+      <div class="field">
+          <label for="profile-username">Username</label>
+          <input type="text" id="profile-username" placeholder="username" value="{{ user.first_name }}" maxlength="30">
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html]: Please [remove](https://www.nngroup.com/articles/form-design-placeholders/) the placeholder attribute.

File: pontoon/contributors/templates/contributors/settings.html [hunk 7]
```diff
@@ -39,6 +34,18 @@ <h2 id="username">
...
+      </div>
+      <div class="field">
+          <label for="profile-email">Email</label>
+          <input type="email" id="profile-email" placeholder="email" value="{{ user.email }}">
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html]: Please [remove](https://www.nngroup.com/articles/form-design-placeholders/) the placeholder attribute.

File: pontoon/contributors/static/css/contributor.css [hunk 8]
```diff
@@ -33,57 +33,42 @@ a.avatar:hover .desc ~ img {
...
-  display: inline-block;
+#profile-form {
+  display: block;
   position: relative;
+  text-align: left;
+  margin: 0 auto 70px auto;
+  width: 620px;
+}
+#profile-form .field {
+  margin-top: 10px;
```
Comment (mathjazz) [inline pontoon/contributors/static/css/contributor.css]: 20px

**Comment by adngdb** [comment]
Hi @karabellyj, do you think you will have time to address the comments from @mathjazz ? If not, please let me know and I can take this over and finish it. Let's not waste your good work. :)

**Comment by jotes** [comment]
@karabellyj @mathjazz @adngdb I'm currently unsure how this feature will work with https://bugzilla.mozilla.org/show_bug.cgi?id=1364676 and if that' patch doesn't require any update in django-allauth. I'm not saying it's wrong somehow, I just didn't have time to look at this from the security standpoint. Besides, :+1: and welcome @karabellyj :-)

**Comment by karabellyj** [comment]
@adngdb Hello, I think I'll be able to finish this PR. If it's time pressing let me know and I'll hand it over to you.

**Comment by karabellyj** [comment]
@mathjazz I added the form validation but a few tests failed, should I adjust?

**Comment by karabellyj** [comment]
@adngdb Hi, I did rebase and think nothing broke. I am struggling with the feature. I don't know where to add code to track email input field change. Can you help?

**Comment by adngdb** [comment]
@karabellyj testing this now, I can't seem to be able to change my email address. It stays the same after I click the "save" button. 

Regarding the 2 requests I made, one of them is simply adding a text stating that changing your email address will mean a log out just next to the input for the address on the settings page. 

The other one requires that, in the settings view, when the user sends new data, you want to compare the new email address with the one in the database. If they are different, by the end of the execution of the view, you will want to log the user out (you can use `request.user.logout()` for that) and redirect to the home page to make sure they have to log in again. 

If you need more help, feel free to reach out to me on IRC (channel #pontoon, nickname adrian).

**Comment by adngdb** [comment]
@karabellyj Are you able to make this work? I run the website using your branch locally, and I see some of your changes, but editing my email address doesn't work (it doesn't change in the database) and I do not get disconnected when I do so (which kinda makes sense since the address didn't change). Any idea what's up? Can you reproduce?

**Comment by adngdb** [comment]
Also, you are breaking some of our PEP8 rules, see this: https://travis-ci.org/mozilla/pontoon/builds/287929686

**Comment by karabellyj** [comment]
@adngdb Okey, I don't have problem with that. I checked and it is saving the address on my localhost. Is it working if you press enter after editing?

File: pontoon/contributors/static/js/contributor.js [hunk 9]
```diff
@@ -27,19 +32,23 @@ $(function() {
...
       save();
     }
   });
+  $('#profile-email').focus(function(e) {
+      e.preventDefault();
+      console.log("Email changed");
+      if($('#profile-email').next('.validation').length == 0) {
+        $('#profile-email').after('<span class=\'validation\' style=\'color:red;margin-left: 10px;\'>Changing email will cause a logout.</span>')
+      }
+  });
```
Comment (adngdb) [inline pontoon/contributors/static/js/contributor.js]: This should be done with CSS instead of JS. I'd put it in a span next to the input and show it on focus using CSS rules.

File: pontoon/contributors/static/js/contributor.js [hunk 10]
```diff
@@ -1,21 +1,26 @@
...
       data: {
         csrfmiddlewaretoken: $('#server').data('csrf'),
-        first_name: $.trim(input.val())
+        first_name: $.trim(username.val()),
+        email: $.trim(email.val())
       },
       success: function(data) {
         if (data === "ok") {
-          input.blur();
           Pontoon.endLoader('Thank you!');
```
Comment (adngdb) [inline pontoon/contributors/static/js/contributor.js]: Do you mind changing this by ``Changes saved!`` instead of this very confusing ``Thank you`` while you're working on this? It's bugging me... :)

File: pontoon/contributors/tests.py [hunk 11]
```diff
@@ -55,7 +56,8 @@ def test_missing_profile_fields(self):
         assert_contains(response, 'This field is required.', count=2, status_code=400)
 
     def test_valid_first_name(self):
-        response = self.client.post('/save-user-profile/', {'first_name': 'contributor', 'email': 'test4@example.com'})
+        response = self.client.post('/save-user-profile/',
+                                    {'first_name': 'contributor', 'email': self.user.email})
```
Comment (adngdb) [inline pontoon/contributors/tests.py]: Please put things on a new line instead, like this:

```python
response = self.client.post(
    '/save-user-profile/',
    {'first_name': 'contributor', 'email': self.user.email}
)
```

File: pontoon/contributors/tests.py [hunk 12]
```diff
@@ -260,9 +262,9 @@ def test_given_period(self):
         Checks if view sets and returns data for right period.
         """
         with patch(
-            'django.utils.timezone.now',
-            wraps=now,
-            return_value=aware_datetime(2015, 7, 5)
+                'django.utils.timezone.now',
+                wraps=now,
+                return_value=aware_datetime(2015, 7, 5)
```
Comment (adngdb) [inline pontoon/contributors/tests.py]: Why did you add that level of indentation? It shouldn't be required. If PEP8 does require it, then instead you should write it like this: 

```python
with (
    patch(
        'django.utils.timezone.now',
        # ...
    )
):
```

File: pontoon/base/forms.py [hunk 13]
```diff
@@ -172,7 +172,8 @@ class Meta:
 
     def clean_email(self):
         email = self.cleaned_data.get('email')
-        if email and User.objects.filter(email=email).exclude(username=self.instance.username).exists():
+        if email and User.objects.filter(email=email)\
+                .exclude(username=self.instance.username).exists():
```
Comment (adngdb) [inline pontoon/base/forms.py]: Please write it like this instead: 

```python
if (
    email and
    (
        User.objects.filter(email=email)
        .exclude(username=self.instance.username)
        .exists()
    )
):
```

All of those comments are for the sake of consistency of coding style across our code base.

File: pontoon/contributors/templates/contributors/settings.html [hunk 14]
```diff
@@ -41,7 +41,6 @@ <h2 id="username">{{ user.first_name }}</h2>
 {% block bottom %}
 <section id="main">
   <div id="profile-form">
-      {% csrf_token %}
```
Comment (adngdb) [inline pontoon/contributors/templates/contributors/settings.html]: Pretty sure that is not the right answer to the problem. Removing the csrf token is going to expose us to attacks, and django doesn't even allow it. 

Sadly, it seems our JS code allows user data editing without requiring the CSRF token. @mathjazz how come that's possible? Isn't that a security issue?

File: pontoon/base/forms.py [hunk 15]
```diff
@@ -159,15 +159,29 @@ def save(self, commit=True):
...
+    email = forms.EmailField()
 
     class Meta:
         model = User
-        fields = ('first_name',)
+        fields = ('first_name', 'email')
+
+    def clean_email(self):
+        email = self.cleaned_data.get('email')
+        if(
```
Comment (adngdb) [inline pontoon/base/forms.py]: You need to add a white space between ``if`` and the parenthesis.

File: pontoon/base/forms.py [hunk 16]
```diff
@@ -159,15 +159,29 @@ def save(self, commit=True):
...
+          email and
```
Comment (adngdb) [inline pontoon/base/forms.py]: This is the wrong level of indentation, you need to add 2 more spaces here and on all other lines of this block. 

Note that you can catch all those errors by running pylama locally. If you're using docker to run Pontoon, here's how you can do it: 

```bash
$ make dockershell
...
$ pylama pontoon
```

File: pontoon/contributors/templates/contributors/settings.html (hunk 14 above)
Comment (adngdb) [reply pontoon/contributors/templates/contributors/settings.html]: Not sure exactly, but I can tell you how to reproduce: 

1. checkout this branch
2. go to your settings page
3. change your email address, and press Enter
4. the change happen, even though no CSRF token is on the page

I am absolutely not sure if this is a problem or not, which is why I'm bringing this up. I assume you've worked on this, so maybe you know?

Comment (mathjazz) [reply pontoon/contributors/templates/contributors/settings.html]: Token is passed here:
https://github.com/mozilla/pontoon/pull/631/files#diff-10c87556f33a2cada9ff50520978ed6dL11

Comment (adngdb) [reply pontoon/contributors/templates/contributors/settings.html]: Aha, OK, the CSRF token is added as a data argument in another HTML tag, perfect. That explains this, and there's no security issue. Thanks, @mathjazz !

**Comment by adngdb** [comment]
Hi @karabellyj, happy new year! As I was reviewing this, I stumbled upon a few more problems, and while trying to figure out what was wrong, realized it would be much easier if I just took over. Hence, I just opened #793, which is basically all your commits (I won't be stealing attribution to your work, don't worry 😉) plus some changes I've made on top of them, to solve the issues I was having during testing. 

Thank you very much for contributing to Pontoon! Looking forward to your next patch! :-)
